modules/sw360/sw360-client/src/test/java/org/eclipse/sw360/antenna/sw360/client/auth/AccessTokenProviderTest.java

```diff
@@ -0,0 +1,166 @@
+/*
+ * Copyright (c) Bosch.IO GmbH 2020.
+ *
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v2.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v20.html
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.eclipse.sw360.antenna.sw360.client.auth;
+
+import org.junit.Before;
+import org.junit.Test;
+
+import java.io.IOException;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionException;
+import java.util.function.Function;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.fail;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+
+public class AccessTokenProviderTest {
+    /**
+     * Constant for a test token.
+     */
+    private static final String TOKEN = "a_test_access_token";
+
+    /**
+     * Mock for the authentication client.
+     */
+    private SW360AuthenticationClient authClient;
+
+    /**
+     * The token provider to be tested.
+     */
+    private AccessTokenProvider tokenProvider;
+
+    @Before
+    public void setUp() {
+        authClient = mock(SW360AuthenticationClient.class);
+        tokenProvider = new AccessTokenProvider(authClient);
+    }
+
+    /**
+     * Prepares the mock for the authentication client to expect requests for
+     * access tokens and sets the results to be returned.
+     *
+     * @param result      the first result
+     * @param moreResults an arbitrary number of further results
+     */
+    @SafeVarargs
+    private final void expectTokenRequest(CompletableFuture<String> result, CompletableFuture<String>... moreResults) {
+        when(authClient.getOAuth2AccessToken()).thenReturn(result, moreResults);
+    }
+
+    @Test
+    public void testAccessTokenCanBeObtained() {
+        CompletableFuture<String> tokenFuture = CompletableFuture.completedFuture(TOKEN);
+        expectTokenRequest(tokenFuture);
+
+        AccessToken accessToken = tokenProvider.obtainAccessToken().join();
+        assertThat(accessToken.getToken()).isEqualTo(TOKEN);
+    }
+
+    @Test
+    public void testAccessTokenIsCached() {
+        CompletableFuture<String> tokenFuture = CompletableFuture.completedFuture(TOKEN);
+        expectTokenRequest(tokenFuture);
+        AccessToken accessToken1 = tokenProvider.obtainAccessToken().join();
+
+        AccessToken accessToken2 = tokenProvider.obtainAccessToken().join();
+        assertThat(accessToken2).isSameAs(accessToken1);
+        verify(authClient).getOAuth2AccessToken();  // exactly once
```

Comment:
I don't know if this comment is necessary, we use this method in this way several time. 
Just a thought though

Reply:
I tend to forget this and interpret such constructs as "verify at least once". Normally I do not use this comment. But the purpose of this test is to check that a cached token is not fetched a second time. Therefore, I prefer to make this explicit.

Reply:
that's fine then :)